pallets/stellar-relay/src/validation.rs

```diff
@@ -0,0 +1,224 @@
+) -> BTreeMap<T::OrganizationId, u32> {
+	let mut validator_count_per_organization_map = BTreeMap::<T::OrganizationId, u32>::new();
+
+	for validator in validators.iter() {
+		validator_count_per_organization_map
+			.entry(validator.organization_id)
+			.and_modify(|e| {
+				*e += 1;
+			})
+			.or_insert(1);
+	}
+
+	validator_count_per_organization_map
+}
+
+/// Builds a map used to identify the targeted organizations
+fn targeted_organization_map<T: Config>(
+	envelopes: &UnlimitedVarArray<ScpEnvelope>,
+	validators: &BoundedVec<ValidatorOf<T>, T::ValidatorLimit>,
+	orgs_length: usize,
+) -> Result<BTreeMap<T::OrganizationId, u32>, Error<T>> {
+	// Find the validators that are targeted by the SCP messages
+	let targeted_validators = validators
+		.iter()
+		.filter(|validator| {
+			envelopes.get_vec().iter().any(|envelope| {
+				envelope.statement.node_id.to_encoding() == validator.public_key.to_vec()
+			})
+		})
+		.collect::<Vec<&ValidatorOf<T>>>();
+
+	// A map is used to avoid duplicates and simultaneously track the number of validators
+	// that were targeted
+	let mut targeted_organization_map = BTreeMap::<T::OrganizationId, u32>::new();
+	for validator in targeted_validators {
+		targeted_organization_map
+			.entry(validator.organization_id)
+			.and_modify(|e| {
+				*e += 1;
+			})
+			.or_insert(1);
+	}
+
+	// Count the number of distinct organizations that are targeted by the SCP messages
+	let targeted_organization_count = targeted_organization_map.len();
+
+	// Check that the distinct organizations occurring in the validator structs related to
+	// the externalized messages are more than 2/3 of the total amount of organizations in
+	// the tier 1 validator set.
+	// Use multiplication to avoid floating point numbers.
+	ensure!(
+		targeted_organization_count * 3 > orgs_length * 2,
+		Error::<T>::InvalidQuorumSetNotEnoughOrganizations
+	);
+
+	Ok(targeted_organization_map)
+}
+
+/// Returns a tuple of Externalized Value and Externalized n_h
+pub fn get_externalized_info<T: Config>(envelope: &ScpEnvelope) -> Result<(&Value, u32), Error<T>> {
+	match &envelope.statement.pledges {
+		ScpStatementPledges::ScpStExternalize(externalized_statement) =>
+			Ok((&externalized_statement.commit.value, externalized_statement.n_h)),
+		ScpStatementPledges::ScpStConfirm(confirmed_statement) =>
+			Ok((&confirmed_statement.ballot.value, confirmed_statement.n_h)),
+		_ => return Err(Error::<T>::InvalidScpPledge),
+	}
+}
+
+/// Returns the node id of the envelope if it is part of the set of validators
+fn get_node_id<T: Config>(
+	envelope: &ScpEnvelope,
+	validators: &BoundedVec<ValidatorOf<T>, T::ValidatorLimit>,
+) -> Result<NodeId, Error<T>> {
+	let node_id = envelope.statement.node_id.clone();
+	let node_id_found = validators
+		.iter()
+		.any(|validator| validator.public_key.to_vec() == node_id.to_encoding());
+
+	ensure!(node_id_found, Error::<T>::EnvelopeSignedByUnknownValidator);
+
+	Ok(node_id)
+}
+
+pub fn check_targets<T: Config>(
```

Comment:
Nit: Maybe we can give this a better name. Maybe `check_quorum_validity` or `check_for_valid_quorum_set`

Reply:
updated